ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -156,100 +156,206 @@ data ValueHandle m = ValueHandle
+    keyBS = toStrictByteString (encKey k)
 
-lmdbInitRangeReadTable :: LedgerConstraint k v
-  => Int
+withSerializationFoldrWithKey ::
+     (k -> v -> b -> b)
+  -> b
+  -> CodecMK k v
   -> LMDB.Database k v
-  -> LMDB.Transaction mode (HD.UtxoValues k v)
-lmdbInitRangeReadTable count db0 =
-  -- This is inadequate. We are folding over the whole table, the
-  -- fiddling with either is to short circuit as best we can.
-  -- TODO improve lmdb-simple bindings to give better access to cursors
-    wrangle <$> LMDB.foldrWithKey go (Right Map.empty) db
+  -> LMDB.Transaction mode b
+withSerializationFoldrWithKey f b (CodecMK _ _ decKey decVal) = foldrWithKey fBS b
   where
-    wrangle = either HD.UtxoValues HD.UtxoValues
-    db = coerceDatabase db0
-    go (LmdbBox k) (LmdbBox v) acc = do
-      m <- acc
-      when (Map.size m >= count) $ Left m
-      pure $ Map.insert k v m
-
-lmdbRangeReadTable :: LedgerConstraint k v
-  => Int
-  -> LMDB.Database k v
-  -> HD.UtxoKeys k v
-  -> LMDB.Transaction mode (HD.UtxoValues k v)
-lmdbRangeReadTable count db0 (HD.UtxoKeys keys) = case Set.lookupMax keys of
-  -- This is inadequate. We are folding over the whole table, the
-  -- fiddling with either is to short circuit as best we can.
-  -- TODO improve lmdb-simple bindings to give better access to cursors
-  Nothing -> pure $ HD.UtxoValues Map.empty
-  Just last_excluded_key ->
-    let
-      db      = coerceDatabase db0
-      wrangle = either HD.UtxoValues HD.UtxoValues
-      go (LmdbBox k) (LmdbBox v) acc
-        | k <= last_excluded_key = acc
-        | otherwise = do
-            m <- acc
-            when (Map.size m >= count) $ Left m
-            pure $ Map.insert k v m
-    in
-      wrangle <$> LMDB.foldrWithKey go (Right Map.empty) db
+    fBS keyBS valBS =
+      f (deserialiseBS "a key" decKey keyBS) (deserialiseBS "a value" decVal valBS)
+
+-- | Deserialise a 'BS.ByteString' using the provided decoder.
+deserialiseBS
+  :: String
+  -- ^ Label to be used for error reporting. This should describe the value to be deserialised.
+  -> (forall s . CBOR.Decoder s a)
+  -> BS.ByteString
+  -> a
+deserialiseBS label decoder bs = either err snd $ deserialiseFromBytes decoder $ LBS.fromStrict bs
+  where
+    err = error $ "withSerializationFoldrWithKey: error deserializing " ++ label ++ " from the database."
 
-lmdbInitTable :: LedgerConstraint k v
-  => String
-  -> LMDB.Database k v
-  -> HD.UtxoValues k v
+{-------------------------------------------------------------------------------
+ Alternatives to LMDB operations that do not rely on Serialise instances
+
+ We cannot (easily and without runtime overhead) satisfy the Serialise
+ constraints that the LMDB.Simple operations require. We have access to the
+ codification and decodification functions provided in CodecMK, thus, we operate
+ directly on ByteStrings.
+
+ TODO: we might want to submit a patch against the upstream LMDB simple package
+ with these new functions.
```

Comment:
I'm not opposed to this idea. What about patching a more abstract interface of the `LMDB Interface that uses CodecMK` section as well? It might be a nice alternative interface to using the `Serialise` typeclass